### ISSUE TITLE
SBF.REM issue

### ISSUE THREAD

**Issue by ancuop** [RedisLabsModules/redablooms#1]
Hi Itamar Haber,

I have tried redablooms module and found that there' s an issue with SBF.REM. 
Please see below commands for my test case.

127.0.0.1:6379> SBF.ADD test k1 +
(integer) 1
127.0.0.1:6379> SBF.CHECK test k1
(integer) 1
**127.0.0.1:6379> SBF.REM test k1 1
(integer) 0**

I understood that reply of SBF.REM should be the number of removed elements so I think SBF.REM should return '1' is it correct?

Cheer,
AnCuop

P/S: in comment "SBF.ADD key elem id" , could you please teach me what is meaning of "id" and how to use it? Thank you very much.

**Comment by itamarhaber** [comment]
Hello @ancuop,

I could not reproduce the above - "proof":

```
127.0.0.1:6379> SBF.ADD test k1 +
(integer) 1
127.0.0.1:6379> SBF.CHECK test k1
(integer) 1
127.0.0.1:6379> SBF.REM test k1 1
(integer) 1
```

Furthermore, another call to `SBF.REM` yields the following in the cli:

```
127.0.0.1:6379> SBF.REM test k1 1
(integer) 1
```

But that is expected, although not handled at the moment (register underflows get printed to stderr until the modules API provides logging facilities - you should be able to see there the following message: "redablooms.so: Error, Decrementing zero").

As for the question in the post scriptum, the best explanation is in [dablooms' README](https://github.com/bitly/dablooms) but I'll try: classic Bloom uses a single bit for each hash of the sample. Counting Bloom uses registers (4-bit in our case) to allow for decrements. However, both "fail" at scaling - once you've initialized the data structure (for a given error & capacity) you're done.

The scaling variant addresses this by "layering" multiple CBFs - each layer is responsible for an id range based on its capacity, and querying for an item is done going through all layers. By increasing the id, scaling can be achieved by adding more layers.

Personally, I'm not too thrilled about this approach and I have Cuckoo Filters on my todo list as an alternative that scales.

**Comment by ancuop** [comment]
Hi @Itamar Haber,

Thank you very much for your explanation about 'id'. I understood. 

About the issue of SBF.REM, I tried to download redabloom and newest redis-unstable in github again and run in 2 steps:

Firstly, I tried with CBF command, this test is correct
redis-cli: 
    127.0.0.1:6379> CBF.ADD pet cat
    (integer) 49
    127.0.0.1:6379> CBF.CHECK pet cat
    (integer) 1
    127.0.0.1:6379> CBF.REM pet cat
    (integer) 48
    127.0.0.1:6379> CBF.CHECK pet cat
    (integer) 0
    127.0.0.1:6379> CBF.REM pet cat
    (integer) 47
redis-server outputs the message:
    redablooms.so: Error, Decrementing zero
    redablooms.so: Error, Decrementing zero
    redablooms.so: Error, Decrementing zero
    redablooms.so: Error, Decrementing zero
    redablooms.so: Error, Decrementing zero

Secondly, I tried with SBF command, 
redis-cli:
    127.0.0.1:6379> SBF.ADD fruit orange +
    (integer) 1
    127.0.0.1:6379> SBF.CHECK fruit orange
    (integer) 1
    127.0.0.1:6379> SBF.REM fruit orange 1
    (integer) 0
    127.0.0.1:6379> SBF.REM fruit orange 1
    (integer) 0
    127.0.0.1:6379> SBF.REM fruit orange 1
    (integer) 0
redis-server:
    Do not output any error message

My environment is Ubuntu 14.04.4 LTS.

**Comment by ancuop** [comment]
In source code of dablooms.c, function new_scaling_bloom() initializes value 0 for header id: 
    cur_bloom->header->id = 0;
However in redabloom, there is no similar command. 

I guess it runs incorrect in my environment because the initial value of cur_bloom->header->id is not 0.

I try to debug source code by add a command in function scaling_bloom_remove() of redabloom

 _for (i = bloom->num_blooms - 1; i >= 0; i--) {
    cur_bloom = bloom->blooms[i];
    // ******\* add print function to debug
    fprintf(stderr, "REM: i=%d, cur_bloom->header->id: %d, id: %d\n", i, cur_bloom->header->id, id);
    if (id >= cur_bloom->header->id) {
      counting_bloom_remove(cur_bloom, s, len);
      return 1;
    }
  }_

After run SBF.REM command, it print:
  REM: cur_bloom->header->id: 14, id: 1

=> That's why SBF.REM always return 0 in my side

**Comment by ancuop** [comment]
I propose to add initialize value for header id at line 607 (redabloom.c). Is it ok?
`cur_bloom->header->id = 0;`
`cur_bloom->header->count = 0;`

**Comment by itamarhaber** [comment]
That's definitely one way to solve this - thanks for tracing it and feel free to make a proper PR :+1: 

However, as I'm brushing up/relearning/discoving C it appears that a common way to avoid the need to initialize is to call `calloc` instead of `malloc` so perhaps it's better to use that here and in a few other places where structs are begotten.

Also worth pointing out is that now that the API provides the alloc family and custom data types, things will need to be revised sooner or later.

**Comment by ancuop** [comment]
Hi @Itamar Haber,
I have created a PR to temporary solve the issue for someone who has the same issue.
I ported the test from dabloom to test again and it runs well on my side

**Comment by itamarhaber** [comment]
Closed by @ancuop with #2
